dagshub/auth/oauth.py

```diff
@@ -0,0 +1,126 @@
+from typing import Optional, Dict
+import appdirs
+import datetime
+import getpass
+import logging
+import os
+import requests
+import traceback
+import urllib
+import uuid
+import yaml
+
+HOST_KEY = "DAGSHUB_CLIENT_HOST"
+DEFAULT_HOST = "https://dagshub.com"
+CACHE_LOCATION_KEY = "DAGSHUB_CLIENT_CONFIG"
+DEFAULT_CACHE_LOCATION = os.path.join(appdirs.user_cache_dir("dagshub"), "tokens")
+CLIENT_ID_KEY = "DAGSHUB_CLIENT_ID"
+DEFAULT_CLIENT_ID = "32b60ba385aa7cecf24046d8195a71c07dd345d9657977863b52e7748e0f0f28"
+
+logger = logging.getLogger(__name__)
+
+
+class OAuthAuthenticator:
+    def __init__(self, **kwargs):
+        self.host: str = kwargs.get("host", os.environ.get(HOST_KEY, DEFAULT_HOST)).strip("/")
+        self.client_id: str = kwargs.get(
+            "client_id", os.environ.get(CLIENT_ID_KEY, DEFAULT_CLIENT_ID)
+        )
+        self.cache_location: str = kwargs.get(
+            "cache_location", os.environ.get(CACHE_LOCATION_KEY, DEFAULT_CACHE_LOCATION)
+        )
+        self._token_cache: Optional[Dict[str, Dict]] = None
```

Comment:
Why this `kwargs` based pattern? Won't it be more convenient to have named args with default values?
I know the default depend on env vars, but those rarely change during runtime, so should be fine to define them at class creation

Reply:
This is purely my laziness because I like chaining .get()s this way, and you can't do that with named args.
It's worse for readability/autocomplete, so I should probably go with named args, you're right

Reply:
Why not actually? Can't you do this:
```python
def foo(bar = os.environ.get('XXX', 'YYY'))
```